### PR TITLE
Experimental async API generation

### PR DESCRIPTION
**Pull request by osa1**
This PR includes #44 (required to be able to use reqwest's API) and #46 (required to be able to build). With this we now generate functions with `_async` suffix that returns a future using reqwest's unstable API.

### CODE REVIEW

**Comment by osa1** [comment]
One thing I forgot to mention is this only works on nightly.

**Comment by dten** [comment]
i'm looking at these, don't mean to be a pain but I'm thinking accept the 1.0 serde pr, then accept the rustfmt bump commit only and do another format then this will need rebasing i guess.. sounds complicated

**Comment by osa1** [comment]
> i'm looking at these, don't mean to be a pain but I'm thinking accept the 1.0 serde pr, then accept the rustfmt bump commit only and do another format then this will need rebasing i guess.. sounds complicated

Changes in the codegen are not too involved so I think it wouldn't be too hard to rebase and re-submit this PR.

> edit: though i see you've basically got those commits too in this pr

Yeah because this won't work without those PRs. So just merging this one would also work I guess.

Note that this won't work without nightly because I used `impl trait` returns. Let me know if this is not OK and I'll try to figure something out.

**Comment by dten** [comment]
the fact that this currently fails to build on stable will need fixing
I think what you've used `impl trait` for should be possible with only stable code

also I'm not happy about unwraps / expects. Haven't looked in enough detail yet to say if there's no alternative

**Comment by osa1** [comment]
Pfff... For some reason async versions are failing during json parse:

```
error: Client(Error { kind: Json(ErrorImpl { code: ExpectedSomeValue, line: 1, column: 1 }), url: None })
```

I'll need to debug this. Let's talk about unwraps/expects afterwards.

**Comment by osa1** [comment]
It works with reqwest HEAD... I'll ask the maintainer to publish the latest version on crates.io.

**Comment by osa1** [comment]
(rebased)

File: src/requests.rs [hunk 1]
```diff
@@ -34,7 +34,7 @@ mod reqwest_support {
 
             let mut response = self.get(url).send()?;
             let mut res_str = String::new();
-            response.read_to_string(&mut res_str).map_err(reqwest::HyperError::from)?;
```
Comment (dten) [inline src/requests.rs]: error handling has turned into a panic, definitely no good

File: Cargo.toml [hunk 2]
```diff
@@ -11,10 +11,13 @@ version = "0.18.0"
 serde = "1.0"
 serde_derive = "1.0"
 serde_json = "1.0"
+futures = "0.1.16"
```
Comment (osa1) [inline Cargo.toml]: This should only be used when `reqwest` feature is enabled. Do you know how to do this @dten ?

Comment (dten) [reply Cargo.toml]: in the toml you make make the dependency optional and define a feature
`unstable = ["futures"]`
then your unstable features should be hidden behind "request" _and_ "unstable"

File: src/requests.rs (hunk 1 above)
Comment (osa1) [reply src/requests.rs]: Right.. I should probably update the `Error` type of this `impl` to include io errors. Not sure how to do that though... (do I have to use something like `error_chain` maybe?)

**Comment by mehcode** [comment]
We seem to be missing `token` here: https://github.com/osa1/slack-rs-api/blob/async_api/codegen/src/generator.rs#L147

**Comment by dten** [comment]
this is unlikely to be merged ever, shall we close it?

**Comment by osa1** [comment]
Yes.

**Pull request closed without merging**